File: src/web/bloom.js
export class BloomFilter {
    constructor(size = 100, hashes = 3) {
        this.size = size;
        this.hashes = hashes;
        this.filter = new Array(size).fill(false);
    }
    
    hash(str, seed) {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            hash = (hash * 31 + str.charCodeAt(i) + seed) % this.size;
        }
        return hash;
    }
    
    add(item) {
        for (let i = 0; i < this.hashes; i++) {
            this.filter[this.hash(item, i)] = true;
        }
    }
    
    test(item) {
        for (let i = 0; i < this.hashes; i++) {
            if (!this.filter[this.hash(item, i)]) return false;
        }
        return true;
    }
}

export default BloomFilter;

File: src/core/src/bloom.rs
use std::collections::HashSet;

pub struct BloomFilter {
    bits: Vec<bool>,
    size: usize,
    hashes: usize,
}

impl BloomFilter {
    pub fn new(size: usize, hashes: usize) -> Self {
        Self {
            bits: vec![false; size],
            size,
            hashes,
        }
    }

    fn hash(&self, item: &str, seed: usize) -> usize {
        let mut hash = 0;
        for (i, byte) in item.bytes().enumerate() {
            hash = hash.wrapping_mul(31).wrapping_add((byte as usize).wrapping_mul(seed + i));
        }
        hash % self.size
    }

    pub fn add(&mut self, item: &str) {
        for i in 0..self.hashes {
            let idx = self.hash(item, i);
            self.bits[idx] = true;
        }
    }

    pub fn contains(&self, item: &str) -> bool {
        for i in 0..self.hashes {
            let idx = self.hash(item, i);
            if !self.bits[idx] {
                return false;
            }
        }
        true
    }
}
